### PR TITLE
[Task 1859435] refactor Database Resource to concrete MySQL and SQL Database Resource

### PR DESCRIPTION
**Pull request by shenqianjin**
[AB#1859435](https://dev.azure.com/mseng/a4d27ce2-a42d-4b71-8eef-78cee9a9728e/_workitems/edit/1859435)

### CODE REVIEW

File: PluginsAndFeatures/azure-toolkit-for-intellij/azure-intellij-resource-connector-lib/src/main/java/com/microsoft/azure/toolkit/intellij/connector/ResourceView.java [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+     */
+    default Icon getIcon() {
+        return null;
+    }
+
+    /**
+     * show properties in file editor.
+     */
+    default void showProperties(Project project) {
+    }
```
Comment (shenqianjin) [inline PluginsAndFeatures/azure-toolkit-for-intellij/azure-intellij-resource-connector-lib/src/main/java/com/microsoft/azure/toolkit/intellij/connector/ResourceView.java:25]: remove, use map and move show prop..

File: PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/azure/toolkit/intellij/connector/database/DatabaseResource.java [hunk 2]
```diff
@@ -57,9 +54,10 @@ public DatabaseResource(@Nonnull String type, @Nonnull final String databaseId)
         this.databaseName = dbId.name();
     }
 
+    /*@Override
     public ResourceId getServerId() {
         return this.serverId;
-    }
+    }*/
```
Comment (shenqianjin) [inline PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/azure/toolkit/intellij/connector/database/DatabaseResource.java:60]: remove

File: PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/azure/toolkit/intellij/connector/database/DatabaseResource.java [hunk 3]
```diff
@@ -74,60 +72,31 @@ public String getDatabaseId() {
...
+    protected static void write(@Nonnull final Element resourceEle, @Nonnull final DatabaseResource resource) {
+        resourceEle.setAttribute(new Attribute(Resource.FIELD_ID, resource.getId()));
+        resourceEle.addContent(new Element("azureResourceId").addContent(resource.getDatabaseId()));
+        resourceEle.addContent(new Element("url").setText(resource.jdbcUrl.toString()));
+        resourceEle.addContent(new Element("username").setText(resource.username));
+        resourceEle.addContent(new Element("passwordSave").setText(resource.password.saveType().name()));
+        if (ArrayUtils.isNotEmpty(resource.password.password())) {
+            PasswordStore.savePassword(resource.getId(), resource.username, resource.password.password(), resource.password.saveType());
         }
+    }
```
Comment (shenqianjin) [inline PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/azure/toolkit/intellij/connector/database/DatabaseResource.java:91]: extract into BaseDatabaseDefinition

File: PluginsAndFeatures/azure-toolkit-for-intellij/azure-intellij-resource-connector-lib/src/main/java/com/microsoft/azure/toolkit/intellij/connector/ResourceView.java (hunk 1 above)
Comment (shenqianjin) [reply PluginsAndFeatures/azure-toolkit-for-intellij/azure-intellij-resource-connector-lib/src/main/java/com/microsoft/azure/toolkit/intellij/connector/ResourceView.java]: updated.

File: PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/azure/toolkit/intellij/connector/database/DatabaseResource.java (hunk 2 above)
Comment (shenqianjin) [reply PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/azure/toolkit/intellij/connector/database/DatabaseResource.java]: removed.

File: PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/azure/toolkit/intellij/connector/database/DatabaseResource.java (hunk 3 above)
Comment (shenqianjin) [reply PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/azure/toolkit/intellij/connector/database/DatabaseResource.java]: extracted.

File: PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/azure/toolkit/intellij/connector/sql/SqlServerDatabaseResource.java [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+import com.microsoft.azure.toolkit.intellij.connector.database.DatabaseResource;
+import lombok.EqualsAndHashCode;
+import lombok.Getter;
+import lombok.RequiredArgsConstructor;
+import org.jdom.Element;
+
+import javax.annotation.Nonnull;
+import javax.annotation.Nullable;
+
+@EqualsAndHashCode(onlyExplicitlyIncluded = true, callSuper = true)
```
Comment (andxu) [inline PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/azure/toolkit/intellij/connector/sql/SqlServerDatabaseResource.java:17]: too similar with MySQLDatabaseResource

File: PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/azure/toolkit/intellij/connector/sql/SqlServerDatabaseResource.java [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+    public SqlServerDatabaseResource(@Nonnull final String serverId, @Nullable final String databaseName) {
+        super(Definition.SQL_SERVER.getType(), serverId, databaseName);
+    }
+
+    public SqlServerDatabaseResource(@Nonnull final String databaseId) {
+        super(Definition.SQL_SERVER.getType(), databaseId);
+    }
+
+    @Override
+    public String getTitle() {
```
Comment (andxu) [inline PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/azure/toolkit/intellij/connector/sql/SqlServerDatabaseResource.java:29]: I would advise 
```java

public String getTitle() {
return "SQL Server";
}

```

File: PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/azure/toolkit/intellij/connector/mysql/ConnectToMySQLAction.java [hunk 6]
```diff
@@ -40,7 +40,7 @@ public AzureIconSymbol getIconSymbol() {
     public void actionPerformed(NodeActionEvent e) {
         final ConnectorDialog<DatabaseResource, ModuleResource> dialog = new ConnectorDialog<>(project);
         final MySqlServer server = this.node.getServer();
-        dialog.setResource(new DatabaseResource(DatabaseResource.Definition.AZURE_MYSQL.getType(), server.id(), null));
+        dialog.setResource(new DatabaseResource(MySQLDatabaseResource.Definition.AZURE_MYSQL.getType(), server.id(), null));
```
Comment (andxu) [inline PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/azure/toolkit/intellij/connector/mysql/ConnectToMySQLAction.java:43]: I like the api like this: MySQLDatabaseResource.from(server.id())

File: PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/azure/toolkit/intellij/connector/mysql/MySQLConnectionDefinitionPreloadingActivity.java [hunk 7]
```diff
@@ -10,17 +10,16 @@
...
 
 public class MySQLConnectionDefinitionPreloadingActivity extends PreloadingActivity {
 
     @Override
     public void preload(@NotNull ProgressIndicator progressIndicator) {
-        final String resourceType = DatabaseResource.Definition.AZURE_MYSQL.getType();
+        final String resourceType = MySQLDatabaseResource.Definition.AZURE_MYSQL.getType();
         final String consumerType = ModuleResource.TYPE;
-        ResourceManager.registerDefinition(DatabaseResource.Definition.AZURE_MYSQL);
+        ResourceManager.registerDefinition(MySQLDatabaseResource.Definition.AZURE_MYSQL);
```
Comment (andxu) [inline PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/azure/toolkit/intellij/connector/mysql/MySQLConnectionDefinitionPreloadingActivity.java:22]: ResourceManager.registerDefinition(MySQLDatabaseResource.AZURE_MYSQL, MySQLDatabaseResource.class);

File: PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/azure/toolkit/intellij/connector/mysql/MySQLSpringDatasourceCompletionContributor.java [hunk 8]
```diff
@@ -21,9 +20,9 @@
     public List<LookupElement> generateLookupElements() {
         List<LookupElement> lookupElements = new ArrayList<>();
         lookupElements.add(LookupElementBuilder
-                .create(DatabaseResource.Definition.AZURE_MYSQL.getType(), "spring.datasource.url")
+                .create(MySQLDatabaseResource.Definition.AZURE_MYSQL.getType(), "spring.datasource.url")
```
Comment (andxu) [inline PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/azure/toolkit/intellij/connector/mysql/MySQLSpringDatasourceCompletionContributor.java:23]: MySQLDatabaseResource.AZURE_MYSQL

**Comment by andxu** [comment]
Will discuss it next week, approve this PR for endgame test.

**Review by andxu — APPROVED** [review]

**Pull request merged**